src/v/kafka/protocol/batch_reader.cc

```diff
@@ -116,4 +116,26 @@ kafka_batch_adapter batch_reader::consume_batch() {
     return kba;
 }
 
+ss::future<batch_reader::storage_t>
+batch_reader::do_load_slice(model::timeout_clock::time_point tp) {
+    using data_t = model::record_batch_reader::data_t;
+    return ss::do_with(data_t{}, [this, tp](data_t& batches) {
+        const auto resources_exceeded = [this, tp] {
+            return is_end_of_stream() || model::timeout_clock::now() > tp;
+        };
+        const auto consume_one = [this, &batches]() {
+            auto kba = consume_batch();
+            if (likely(kba.v2_format && kba.valid_crc && kba.batch)) {
+                batches.push_back(std::move(*kba.batch));
```

Comment:
This is pretty specialized for the use case of parsing kafka batches off the wire in the context of request handling, so these special cases with magic and crc checking aren't a big deal in the short term.

I think Ben had the idea of adding magic to the in-memory model::record_batch so we could use that to also represent kafka batches off the wire. I liked that, but now I'm realizing this is all a special case of `storage::batch_consumer` but instead of reading our format off the disk, its kafka format from an iobuf.

It would basically look something like

```
class batch_consumer {
virtual consume_result consume_batch_start = 0;
virtual void consume_records(iobuf&&) = 0;
virtual stop_parser consume_batch_end() = 0;
};

class continuous_batch_parser {
continuous_batch_parser(
  std::unique_ptr<batch_consumer> consumer,
  iobuf data);

// continues to parse until stop_parser is reached or end of stream
ss::future<result<size_t>> consume();
```

So in this case the kafka batch adapter used with produce request could do its own crc checking and magic verifications however it wanted. The client will probably do something else, too.

Then the way to compose this with the rest of the system is to wrap a parser/consumer in a record batch reader.

@senior7515 thoughts? seems fully analgous to disk log parser with the record batch reader interface layered on top.